### ISSUE TITLE
User request: add solute Sodium Chloride

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/beers-law-lab#169]
Relocating this issue from https://github.com/phetsims/concentration/issues/49.

Specifications in https://github.com/phetsims/concentration/issues/49#issuecomment-215490920
> Common name: Table salt
formula: NaCl
stock concentration: 5.50 mol/L
molar mass: 58.443 g/mol
solubility: 6.15 mol/L
3 colors:
min: no color-fully transparent
middle: no color-fully transparent
max: Would be nice to add a bit of opacity here. Can we try f1f2f2 at 40% transparency?
color for precipitate particles: f1f2f2 (very light gray)

Remember to add to beers-law-lab-api.js.

**Comment by pixelzoom** [comment]
@amyh-phet where should NaCl be in this list? The list is currently arranged in rainbow (ROYGBIV) order.

![screenshot_30](https://cloud.githubusercontent.com/assets/3046552/15226671/ee4a417c-1840-11e6-84d9-63db0c5c5d9b.png)

**Comment by pixelzoom** [comment]
@amyh-phet ... and are we sure that we want it to appear as "Table salt"?  There are other "chloride" solutes in the list.

**Comment by amyh-phet** [comment]
@pixelzoom  I would suggest we place sodium chloride at the bottom of the list. This would maintain the ROYGBIV order.  I would prefer to call it sodium chloride rather than table salt.

**Comment by pixelzoom** [comment]
When I asked for colors in https://github.com/phetsims/concentration/issues/49#issuecomment-215234764, I said:
> • 3 colors (min, middle, max) used to denote solution concentration (all fully transparent?)
• color for precipitate particles (fully transparent?)

My request was incorrect. For solute color scheme, I need 3 opaque colors, and I need concentrations associated with those 3 colors. For example, for "Copper sulfate", I currently have:

```js
var waterColor = new Color(  224, 255, 255  );
var copperSulfateColorScheme = new SoluteColorScheme( 
  0, waterColor,  // min concentration & color
  0.2, new Color( 200, 225, 255 ), // middle concentration & color
  1.38, new Color( 30, 144, 255 ) // max concentration & color
);
```

@amyh-phet Can you please specify 3 concentrations and colors that we can start with?  I'm guessing that for 'min', the concentration will be 0, and associated color will be waterColor.

**Comment by amyh-phet** [comment]
Fill in the blanks:

min concentration = 0
min color = 224, 255, 255
middle concentration = 3.10
middle color =224, 255, 255
max concentration = 6.15
max color = 224, 255, 255

@pixelzoom  Sodium chloride solutions typically do not look different than water.   I think it might work best to just keep the solution the same as water to emphasize this point.

**Comment by pixelzoom** [comment]
@amyh-phet here's a dev version for review:
http://www.colorado.edu/physics/phet/dev/html/beers-law-lab/1.4.0-dev.7/beers-law-lab_en.html

If this looks OK, assign it back to me so that I can integrate into a maintenance release for both Beer's Law Lab and Concentration.

**Comment by amyh-phet** [comment]
Looks good.

**Comment by pixelzoom** [comment]
@samreid In case you have any need to deploy Beer's Law Lab or Concentration for PhET-IO, please note that this issue is now "in progress".

**Comment by pixelzoom** [comment]
In https://github.com/phetsims/concentration/issues/49, I asked:
> Maintenance release or minor version increment? A maintenance release will have higher development cost, but require less testing. A minor release will have lower development cost, but requires a full RC test cycle, and requires triaging open issues to see what other changes will be picked up.

@ariel-phet replied:
> Definitely a minor release.

@ariel-phet I will defer publication of this until you want to schedule it. But note that this may impact publication of PhET-iO versions of this sim.

**Comment by pixelzoom** [comment]
@samreid Can you please double check my changes to beers-law-lab-api.js in @https://github.com/phetsims/phet-io/commit/c1ad17010df50c5fa8a2dd0644cca909fe35783e? Nothing seems amiss, but I'm not entirely confident that this is the only change required to add a new TSolute.

**Comment by samreid** [comment]
That's the only change the API file needs.

**Comment by pixelzoom** [comment]
@samreid To inform the scheduling of RC testing, can you please cmment on:
• Will having this issue "in progress" would block PhET-iO publication of Beer's Law Lab and Concentration?
• Are there any such publications on the horizon?

**Comment by samreid** [comment]
> • Will having this issue "in progress" block PhET-iO publication of Beer's Law Lab and Concentration?

Does "in progress" imply inconsistent/broken/unusable?  If so, it could cause problems.

> Are there any such publications on the horizon?

One client requested #172 to be fixed, so that may be deployed soon, though I proposed a workaround they are evaluating.  Also, it was proposed that BLL be tested as part of https://github.com/phetsims/phet-io/issues/440 but we have switched to Capacitor Lab: Basics for that test.

**Comment by samreid** [comment]
Oh, the other issue is that there is a branch of beers-law-lab that is destined for master.  We'll have to merge changes from this issue.

**Comment by pixelzoom** [comment]
> Does "in progress" imply inconsistent/broken/unusable? If so, it could cause problems.

In this specific case, "in progress" means that there's a new feature (addition of NaCl solute) that must go through RC testing before publication to the PhET website. I don't believe that anything has been broken by adding this feature, but no promises.

> One client requested #172 to be fixed, so that may be deployed soon

If you deploy from master, you're going to pick up this new feature. And you may want to alert the client to the fact that there's a new solute (NaCl) in the Concentration screen.

**Comment by pixelzoom** [comment]
> Oh, the other issue is that there is a branch of beers-law-lab that is destined for master. We'll have to merge changes from this issue.

Should that merge be done before or after RC testing for public release?

**Comment by samreid** [comment]
> Should that merge be done before or after RC testing for public release?

None of the phet-io changes are necessary for the next public release of Beer's Law Lab, so I'd recommend creating a release branch when you are ready.  The chipper/joist/tandem/phet-io/beers-law-lab branches aren't merging into master until they pass testing in https://github.com/phetsims/function-builder/issues/42 and https://github.com/phetsims/phet-io/issues/440 and I don't know the timeline for that testing (perhaps @phet-steele could comment regarding anticipated timeline).

**Comment by pixelzoom** [comment]
For regression test, please verify that:
• Sodium chloride appears last in the combo box in the Concentration screen
• Sodium chloride does not appear in the Beer's Law screen
• The shaker and dropper on the Concentration screen are labeled 'NaCl'

**Comment by alynne-phet** [comment]
@pixelzoom All of the above points have been verified on the ipads and Mac

**Comment by pixelzoom** [comment]
Additional thoughts...

(1) We should confirm that 6.15 moles is indeed the saturation point for NaCl. When gathering specifications, I'm not sure if it was clear that "max concentration" was the concentration at which precipitation occurs.

(2) I could investigate increasing the amount of solute in the shaker and dropper, but I'm not sure what other problems that might cause.

**Comment by pixelzoom** [comment]
Reported by @brroberts1231 in https://github.com/phetsims/beers-law-lab/issues/169, moving here:

Test device: Schrodinger

Operating System: Win 8

Browser: Chrome

Problem description: When the container has the full 1 Liter of water in it, the shaker will sometimes stop dispensing NaCl before the solution becomes saturated. This doesn't happen with any other solute. 

Steps to reproduce: Select NaCl, fill the container full with water, and shake until no more comes out of the shaker. Most likely, on the first try, it will stop dispensing before it becomes saturated (saturates at 6.150 mol/L). If it doesn't happen on the first try, draining some solution, then adding water and shaking again may cause it to stop before it is saturated. But, this method sometimes also fixes the problem. If it stopped dispensing before saturation, and you drain, refill, and shake again, it may become saturated.

Severity: Low

Screenshots:
![concentration](https://cloud.githubusercontent.com/assets/11947677/15480780/80d5f328-20e4-11e6-9616-a4aa2e691e86.gif)


Troubleshooting information (do not edit):
Name: ‪Beer's Law Lab‬
URL: http://www.colorado.edu/physics/phet/dev/html/beers-law-lab/1.4.0-rc.1/beers-law-lab_en.html
Version: 1.4.0-rc.1 2016-05-23 15:41:35 UTC
Features missing: touch
User Agent: Mozilla/5.0 (Windows NT 6.3; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.102 Safari/537.36
Language: en-US
Window: 1920x973
Pixel Ratio: 1/1
WebGL: WebGL 1.0 (OpenGL ES 2.0 Chromium)
GLSL: WebGL GLSL ES 1.0 (OpenGL ES GLSL ES 1.0 Chromium)
Vendor: WebKit (WebKit WebGL)
Vertex: attribs: 16 varying: 30 uniform: 1024
Texture: size: 8192 imageUnits: 16 (vertex: 16, combined: 32)
Max viewport: 8192x8192
OES_texture_float: true
Dependencies JSON: {"assert":{"sha":"7d27130a","branch":"master"},"axon":{"sha":"9bf8f88c","branch":"master"},"babel":{"sha":"37077fff","branch":"master"},"beers-law-lab":{"sha":"27ef4131","branch":"1.4"},"brand":{"sha":"bb1cfcc6","branch":"master"},"chipper":{"sha":"22a9ffe2","branch":"master"},"dot":{"sha":"04efe0bc","branch":"master"},"joist":{"sha":"3931eb76","branch":"master"},"kite":{"sha":"dce9765e","branch":"master"},"nitroglycerin":{"sha":"ad3cb990","branch":"master"},"phet-core":{"sha":"c48bf320","branch":"master"},"phetcommon":{"sha":"eab7fb7f","branch":"master"},"scenery":{"sha":"d36b3155","branch":"master"},"scenery-phet":{"sha":"2b60b8e4","branch":"master"},"sherpa":{"sha":"e1b3c84b","branch":"master"},"sun":{"sha":"6f66368f","branch":"master"},"tandem":{"sha":"29f64233","branch":"master"}}

related to phetsims/tasks#600 and phetsims/tasks#602

**Comment by pixelzoom** [comment]
@ariel-phet said:
> @pixelzoom marking as low priorty since obviously function builder takes precedence and this is a feature enhancement and not a bug in the published sim.

**Comment by pixelzoom** [comment]
@alynne-phet said:
> @pixelzoom @brroberts1231 Similar issue seen on Ipad 2 (IOS 8.4.1). Shaker completely stopped dispensing NaCl when the solution was not saturated (stopped dispensing at 6.055 mol/L).

**Comment by pixelzoom** [comment]
The issue is that the sim model is only capable of dispensing up to 6 moles of solute from the shaker or beaker. And according to the specification provided by @amyh-phet in #169 (comment), saturation occurs at the maximum concentration of 6.15 moles. So within the limitations of the current model, it's impossible to reach saturation before running out of solute in the shaker or dropper.

Some thoughts...

(1) We should confirm that 6.15 moles is indeed the saturation point for NaCl. When gathering specifications, I'm not sure if it was clear that "max concentration" was the concentration at which precipitation occurs.

(2) I could investigate increasing the amount of solute in the shaker and dropper, but I'm not sure what other problems that might cause. (Note to self: see SOLUTE_AMOUNT_RANGE in ConcentrationModel.)

@amyh-phet how do you want to proceed?

**Comment by pixelzoom** [comment]
A couple of observations....

(1) With less solution in the beaker, NaCl will precipitate.  So I don't think we necessarily have a problem here.

(2) Even with a full 1L, I'm observing some precipitation.  Which is odd, since the max amount of solute should be 6 mol.  That may be a different problem.

**Comment by phet-steele** [comment]
Different saturation procedures yield different results, regarding amount of visible precipitate:
1. Drain beaker completely.
2. Exhaust the NaCl shaker.
3. Fill solvent to 1L.
![trial1](https://cloud.githubusercontent.com/assets/5863899/15485239/6c8c1dd6-20fb-11e6-9f04-b216e1fa7414.PNG)

1. Fill beaker completely.
2. Exhaust NaCl shaker.
![trial2](https://cloud.githubusercontent.com/assets/5863899/15485242/70e01b3a-20fb-11e6-83e9-6f9d4dadc449.PNG)

**Comment by amyh-phet** [comment]
@pixelzoom  The saturation point for sodium chloride is indeed 6.15 mol/L.  I used the same CRC Handbook as we used for saturation points of the other solutes in the simulation. Concentrations above 6.15 mol/L should result in saturation of NaCl.

**Comment by pixelzoom** [comment]
The amount of solute dispensed by the shaker is exceeding 6 moles, and that shouldn't happen. Tracking that issue in https://github.com/phetsims/beers-law-lab/issues/176.

**Comment by pixelzoom** [comment]
Moved new issues to #174, #177, #178.

No issues remaining that are specific to NaCl.  Closing.

**Comment by pixelzoom** [comment]
Actually... I'll leave this open and labeled "in progress" until 1.4.0 is published, in case there's any other need to publish (eg, PhET-iO).

**Comment by pixelzoom** [comment]
Published in Beer's Law Lab 1.4.0 and Concentration 1.3.0.
